paper/src/main/java/com/programmerdan/minecraft/simpleadminhacks/hacks/SaplingHack.java

```diff
@@ -0,0 +1,95 @@
+package com.programmerdan.minecraft.simpleadminhacks.hacks;
+
+import com.programmerdan.minecraft.simpleadminhacks.SimpleAdminHacks;
+import com.programmerdan.minecraft.simpleadminhacks.configs.SaplingConfig;
+import com.programmerdan.minecraft.simpleadminhacks.framework.SimpleHack;
+import java.util.Random;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockBreakEvent;
+import org.bukkit.event.block.LeavesDecayEvent;
+import org.bukkit.inventory.ItemStack;
+
+public class SaplingHack extends SimpleHack<SaplingConfig> implements Listener {
+
+	private Random random = new Random();
+
+	public SaplingHack(SimpleAdminHacks plugin, SaplingConfig config) {
+		super(plugin, config);
+	}
+
+	@Override
+	public void onEnable() {
+		super.onEnable();
+		this.plugin.registerListener(this);
+	}
+
+	@EventHandler
+	public void onBlockBreak(BlockBreakEvent event) {
+		Block brokenBlock = event.getBlock();
+		if (!checkIfLeafBlock(brokenBlock)) {
+			return;
+		}
+		if (!this.config.isAllowFortune()) {
+			if (checkIfItemHasFortune(event.getPlayer().getItemInUse())) {
+				return;
+			}
+		}
+		if (!this.config.getChanceMap().containsKey(brokenBlock.getType())) {
+			return;
+		}
+		double nextDouble = random.nextDouble();
+		if (nextDouble <= this.config.getChanceMap().get(brokenBlock.getType())) {
+			brokenBlock.getWorld().dropItemNaturally(brokenBlock.getLocation(), new ItemStack(getSaplingMaterial(brokenBlock.getType()), 1));
+			logger.info("Dropped extra " + getSaplingMaterial(brokenBlock.getType())  + " from being broken at " + brokenBlock.getLocation());
+		}
+	}
+
+	@EventHandler
+	public void onLeafDecay(LeavesDecayEvent event) {
+		Block decayedBlock = event.getBlock();
+		if (!checkIfLeafBlock(decayedBlock)) {
+			return;
+		}
+		double nextDouble = random.nextDouble();
+		if (nextDouble <= this.config.getChanceMap().get(decayedBlock.getType())) {
+			decayedBlock.getWorld().dropItemNaturally(decayedBlock.getLocation(), new ItemStack(getSaplingMaterial(decayedBlock.getType()), 1));
+			logger.info("Dropped extra " + getSaplingMaterial(decayedBlock.getType())  + " from being broken at " + decayedBlock.getLocation());
```

Comment:
```java
final Material dropMaterial = getSaplingMaterial(decayedBlock.getType());
if (dropMaterial != Material.AIR) {
	final Location dropLocation = decayedBlock.getLocation();
	decayedBlock.getWorld().dropItemNaturally(dropLocation, new ItemStack(dropMaterial, 1));
	logger.info("Dropped extra " + dropMaterial + " from being broken at " + dropLocation);
}
```
This code would avoid the unnecessary:
- Instantiation of an AIR ItemStack
- Instantiation (and insertion into the world) of an entity representing that AIR ItemStack
- Logging to the console of the dropping of that AIR ItemStack

Reply:
I would leave it as is.

The only thing I can suggest in this part of the code.
**onBlockBreak** and **onLeafDecay** use the same code to generate random saplings, this can be moved to a separate function.

The code, in general, can be refactored though I would suggest not making any changes for now and just accept it as is and make refactoring later.

Reply:
Agreed on the code duplication point.